[dxva] Fix flash/wrong picture when skipping

CrystalP opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

When the dxva decoder is on a separate d3d device (gpu of the last 5 years, maybe more), the GPU doesn't always wait for the decoded images to be ready before they're rendered. The same thing happens with both shared and copy buffers.

There is no issue during normal playback but when skipping, wrong pictures or white pictures are briefly displayed, probably because resuming playback from a new position requires more work than the time available during a frame. Maybe it wasn't a problem for 1080p and became visible with 4k increased resource use. Problem only noticed by me on AMD with 4k hevc at this time.
edit: confirmed on nVidia RTX 4070 with software renderer as well

D3D11 provides two synchronization methods: ID3D11Query (any D3D11 system) and ID3D11Fence (Windows 10 Creators Update and above). Only D3D11Fence worked for AMD (unknown for nVidia) and no problem was reported in Windows 8, so there was no reason to implement both.

The concept is the same for query and fence:
- when a picture is decoded (buffer Initialize() function), mark the point in time in the GPU command queue (new function SetFence())
- when the rendering code retrieves the picture (buffer GetResource() function), wait until the point in time marked above is reached by the GPU, to be sure that the picture data is valid before using it any further.
- in addition, during the first use of a buffer (Initialize()), perform various initializations (new function InitializeFence()).
If for any reason the one-time initialization is unsuccessful, the code will revert to pre-PR behavior.
- added destructor to release system handles

Two variations were possible (make the CPU wait, or make the GPU wait without CPU interruption). GPU wait is better in theory, in practice no difference observed (visual or in Task Manager, Visual Studio Performance Profile), maybe due to cpu and gpu idle most of the time. CPU wait degraded things a little bit on Intel 13th gen (which doesn't need the wait at all...) with a tiny freeze after a seek. It seemed best to keep only the GPU wait implementation.

After testing, this code doesn't allow hardware that needed the buffer copy to work without (ie nVidia drivers < 465, AMD prior to GCN 5) so activation conditions of shared surfaces were not modified.

New fence code activated only for AMD and nVidia which benefit from it, and for Window 10 Creators Update and above only.

Notes:
* possibility of UINT64 fenceValue wrapping after reaching UINT64_MAX: tested and the fence doesn't react well, but it won't happen for hundreds of years of continuous playback when the fence starts at value 0. In other words, it's not a problem :-)
* there is no need for fences in other parts of the code so I didn't separate the fence code into its own class, but it wouldn't be much effort.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Self-reported issue #23518, most UHD BD titles have flashes / wrong picture displayed when skipping. Seems limited to AMD and 4k (heavier load on the decoder probably)

Then a comment mentioned that this may fix the shared buffer path for GPUs that need the copy path in master. GPU and memory usage would be reduced for those, but testing is required.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Windows 10, AMD 5600G, Intel 8th and 13th gen
* Skip problem
Played the titles that had issues before (including the sample) => problem removed
Tested other files that didn't have problems => still don't
Tried samples of various accelerated codecs (wmv, mpeg2, avc, hevc). I don't have av1.

Compared cpu/gpu/memory utilization before and after: no clear difference found.

I used to see half-complete frames from time to time during normal playback on AMD and it's not happening anymore.

Testing requested in forum and team slack for previous gen GPU.

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->
Artifact-free skipping for UHD BD on AMD using default options, and nVidia in more limited situations (software renderer)
Maybe it also helps from time to time during normal playback when the system is loaded.

## Screenshots (if appropriate):
Glitch:
https://github.com/xbmc/xbmc/assets/489377/7bc82a0f-a8c5-48f9-921f-17ecbad3adec

no glitch with PR:

https://github.com/xbmc/xbmc/assets/489377/a17b51cd-86e0-487e-aec3-27f339c4d84a


## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

CrystalP committed:
[dxva] Add synchronization for the separate decoder device / shared buffers path

This fixes flashes or wrong picture when skipping in videos for AMD/all render methods
and nVidia/software renderer. No issue with Intel.
The issue is due to the lack of synchronization between the main d3d device and the device
of the decoder (when separate). Video decoding completes after the surface has already been
rendered, maybe even presented.

ID3D11Fence, which fixes both AMD and nVidia, is available on Windows 10 Creators Update and above only.
ID3D11Query doesn't help with AMD, unknown status for nVidia.

The GPU fence is set right after the picture decode is queued and a GPU wait instruction is placed
in the queue of the device used for rendering.

thexai commented:
Quick tested on NVIDIA RTX 4070 and not found regressions when "true shared" path is used but video not feels smooth when used copy path. 

More tests are needed (also on Intel) but it would probably be safer to limit changes to "true shared" path only, to not cause regressions on old systems. Keep in mind that it improves issues in seek only and there is a potential risk of creating much bigger problems in playback.

Another option is to enable this change for AMD only for now and wait for feedback.

CrystalP commented:
I'm also concerned about the older gpu. I'll try to prepare some test builds to request feedback in the forum, one with CPU wait one with GPU wait.

Intel 13 felt fine to me but I don't have a 4k sample with a long pan to test smoothness, only 720p or 1080p, which don't stress the GPU as much but were ok.

The fix is needed for both paths for recent AMD like mine (and both felt smooth) for skipping, but I also suspect the occasional dropped frame / half-rendered frame during normal playback may have the same root cause. I'll be dog-fooding this to look for differences.

Maybe your nVidia would have different result with cpu wait (1+3) and gpu wait (1+2+3).

CrystalP commented:
Win-64 test builds:
cpuwait http://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20230724-fdab2179-fix-flash-cpuwait-x64.exe
gpuwait http://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20230724-a4876479-fix-flash-x64.exe

thexai commented:
Runtime tested on Intel NUC8i3BEK - Windows 11 22H2

Both versions causes noticeable micro stuttering all the time but more noticeable when playing 4K HDR with HDR off (tone mapping) probably because GPU has more load. Also bit stutters when enabling debug OSD (not happens before in master).

Seems this causes severe performance degradation in Intel probably because built-in Intel synchronization routines are much more optimized. Intel always has supported "true shared" path in all drivers.

> The fix is needed for both paths for recent AMD like mine

copy texture path is entirely a workaround and is a "dead road". I meant is no make sense use when new drivers (NVIDIA and AMD) supports true shared path. It is not worth putting effort into it.

Also the only reason exist is "old" drivers has synchronization issues without copy. If changes in "true shared" path improves synchronization in AMD and NVIDIA is probably old drivers becomes usable in now and copy path could be removed entirely or restrict only to "Windows 8 + old drivers".

My recommendations: 

- Apply changes only for NVIDIA and AMD in true shared path. 
- Copy texture path is a dead path (no worth make changes).
- Re-test if old NVIDIA drivers are usable with this PR in true shared path
- Slowly kill copy path:

```c++
  m_DXVA2SharedDecoderSurfaces =
      ad.VendorId == PCIV_Intel ||
      (ad.VendorId == PCIV_NVIDIA && driver.valid && driver.majorVersion >= 465) ||
      (ad.VendorId == PCIV_AMD && driver.valid && driver.majorVersion >= 30 &&
       m_d3dFeatureLevel >= D3D_FEATURE_LEVEL_12_1) ||
       Windows >= Windows_10_Creators_Update;
```

CrystalP commented:
I'll remove the changes to the copy path.

I didn't know the reason for the copy path was the synchronization. Then it would indeed be interesting to find out if there are improvements for older drivers and gpu, and it would be a very positive change.

I haven't tried much HDR yet, need to locate a good sequence to evaluate smoothness.

Interesting results with the nuc, I don't have access to an Intel 8th gen / Windows 10 anymore to clarify if it's about the generation or the OS. The 13th gen is fine on Windows 10, but it could be just because it is more powerful.

CrystalP committed:
[Windows] define Creators Update Windows version

CrystalP commented:
PR description and commit messages will have to be rewritten if this turns out to help older gpus and drivers.

Reverted changes to the buffer copy path, restored the Flush()

Updated the conditions to enable shared surfaces:
- Intel GPU
or
- Windows 10 Creators Update (1703) + gpu/driver with feature level 11_0
FL 11_0 is a guess: D3D11Fence is the same as D3D12Fence per MS doc, and D3D12 requires a minimum FL 11_0. Maybe it will have to be bumped to 12_0, or 12_1 like current master, we'll see the testing results.

The fences will be created and used when the correct objects can be created (should always be possible for Windows 1703 and up) - to be adjusted based on feedback.
For testing purposes Intel will also use the fence sync, to receive additional feedback. It is most likely that the fence sync will be disabled for Intel for the merge, since it already works fine without.

About NUC / Intel 8th gen: would a Flush() help maybe? Maybe in GetResource() before context4->Wait(), or in SetFence() after m_deviceContext4->Signal().

thexai commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/DXVA.cpp:
```diff
@@ -807,7 +849,79 @@ void CVideoBufferShared::Initialize(CDecoder* decoder)
   if (handle == INVALID_HANDLE_VALUE)
+  {
     handle = decoder->m_sharedHandle;
+    InitializeFence(decoder);
```
As mentioned in comments this ONLY should be initialized for AMD and NVIDIA

thexai commented on xbmc/rendering/dx/DeviceResources.cpp:
```diff
@@ -1156,9 +1156,8 @@ void DX::DeviceResources::CheckDXVA2SharedDecoderSurfaces()
 
   m_DXVA2SharedDecoderSurfaces =
       ad.VendorId == PCIV_Intel ||
-      (ad.VendorId == PCIV_NVIDIA && driver.valid && driver.majorVersion >= 465) ||
```
Why remove NVIDIA driver condition?

Nvidia driver > 465 supports true shared path without any other condition i.e Windows 10 < Creators Update and without this PR (fence).

Also `D3D_FEATURE_LEVEL_12_1` is relevant only for AMD because identifies HW gen that supports true shared path.

Please change to already commented:

```
  m_DXVA2SharedDecoderSurfaces =
      ad.VendorId == PCIV_Intel ||
      (ad.VendorId == PCIV_NVIDIA && driver.valid && driver.majorVersion >= 465) ||
      (ad.VendorId == PCIV_AMD && driver.valid && driver.majorVersion >= 30 &&
       m_d3dFeatureLevel >= D3D_FEATURE_LEVEL_12_1) ||
      CSysInfo::IsWindowsVersionAtLeast(CSysInfo::WindowsVersionWin10_1703));
```

Anyway this change should go in with a separate commit or follow-up PR because we must first confirm that it works for Nvidia and AMD. Maybe it only works for Nvidia and then it will have to be modified.

CrystalP replied:
This is for testing purposes with a special build at this point, so things are wide open to activate for a large number of systems on purpose.
The merge button is not on fire and I had no intention to merge as is.
I'm publishing a couple test builds in the forum to receive feedback on what works and what doesn't.
We can then discuss and agree on the conditions before a merge.

CrystalP replied:
... but I didn't account for the fact that maybe nVidia and/or AMD had a working shared path under Windows 8.
Did they?

thexai replied:
Yes, Nvidia with drivers >= 465 should work since no one has complained.

AMD new HW (which is reported with DirectX FL 12.1) should work but since Windows 8 is limited to FL 11.1, this path is not used.

CrystalP replied:
ok so let's not open that door, unless someone shows up with a computer that allows testing.

CrystalP commented:
> > About NUC / Intel 8th gen: would a Flush() help maybe?
> 
> No, as already commented Intel NUC and Intel in general NOT uses copy path, ALWAYS use true shared that not contains any Flush(). Also intel not need "help" because works perfectly fine in master/Nexus (any changes you make can only make things worse).

No problem, it bugs me when something that should work doesn't. But the problem may lie elsewhere, looking in Visual Studio Performance Profiler shows a GPU pattern that seems strange to me. So far I haven't been able to run it with other players for comparison.

thexai commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/DXVA.cpp:
```diff
@@ -799,6 +804,43 @@ HRESULT CVideoBufferShared::GetResource(ID3D11Resource** ppResource)
+  {
+    // Make the GPU wait for the fence value that produced the picture
+    ComPtr<ID3D11DeviceContext1> context1 = DX::DeviceResources::Get()->GetImmediateContext();
+    ComPtr<ID3D11DeviceContext4> context4;
```
Is not necessary obtain this interface for every frame during playback. This code block can be moved inside previous if block (executed only at initialization):

` if (!m_sharedFence && m_handleFence != INVALID_HANDLE_VALUE)`

and save `ID3D11DeviceContext4` as member variable.

Then in every frame is enough execute `m_context4->Wait` line

thexai commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/DXVA.cpp:
```diff
@@ -793,21 +798,141 @@ HRESULT CVideoBufferShared::GetResource(ID3D11Resource** ppResource)
+    }
+  }
+
+  if (FAILED(hr = m_sharedRes.CopyTo(ppResource)))
```
Now seems more logic move this code at end of method (when all is initialized) and after "wait"

```c++
hr = m_sharedRes.CopyTo(ppResource);

return hr;
```

CrystalP replied:
it's a pointer copy, but OK.

moving the fence initialization in the if (m_sharedRes) block since there is no need to log errors for each picture if something is wrong in the init.

thexai commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/DXVA.cpp:
```diff
@@ -793,21 +798,141 @@ HRESULT CVideoBufferShared::GetResource(ID3D11Resource** ppResource)
+  {
+    // Make the GPU wait for the fence value that produced the picture
+    ComPtr<ID3D11DeviceContext1> context1 = DX::DeviceResources::Get()->GetImmediateContext();
+    ComPtr<ID3D11DeviceContext4> context4;
```
`context1` and `context4` are unused variables now

CrystalP replied:
yes, multitasking is no good. I'll get there.

enen92 commented:
Gave it a go but in the hopes of fixing a small green "flash" I noticed while testing the libbluray bump but unfortunately this is giving me completely borked results for me:

https://github.com/xbmc/xbmc/assets/7375276/d3c72b94-1ad2-46af-b7a8-c7980b1b7b58

Current master where you can see a small green flash when seeking (but generally works well):

https://github.com/xbmc/xbmc/assets/7375276/59802a69-3e10-4783-a5e2-5389e8376702

This is with an AMD Radeon RX550

CrystalP commented:
lol that's terrible... Thanks for the testing. There is not much hope for the AMD older than yours then.
I remember that green flash from a much older ATI HD4550. A Clear() in the right place may help but the old card won't run in my modern-ish PC unfortunately.
Maybe the older ID3D11Query would work better for the RX550 and I set some code aside during development. Maybe I'll trouble you for a test after this PR is done.

thexai commented:
Based on these tests/conclusions https://github.com/xbmc/xbmc/issues/23518#issuecomment-1656693436 is not necessary change zero-copy enabling conditions in DeviceResources and only is need enable fence code (GPU wait version) for AMD and NVIDIA (recent HW/drivers).

Also for enable fence code I suggest move logic to DeviceResources like:

```c++
  m_DXVA2UseFence =
      m_DXVA2SharedDecoderSurfaces && (ad.VendorId == PCIV_NVIDIA || ad.VendorId == PCIV_AMD);
```

And in DXVA.cpp only:

```c++
    if (DX::DeviceResources::Get()->UseFence())
      InitializeFence(decoder);
```

With this I think this PR is almost ready  🙂

CrystalP commented:
@enen92 what is the driver version for the RX550? Is that the latest available?

thexai reviewed (approved):
This is a substantial improvement on the main DXVA2 code path today.

enen92 commented:
> This is a substantial improvement on the main DXVA2 code path today.

Not near the computer today but I'll let you know (and runtime test again) tomorrow

thexai commented:
RX550 is unaffected by this change because not uses zero-copy path (AMD DirectX FL < 12.1) then will behave same as before

CrystalP commented:
I asked just in case your drivers were not up to date, then the result may be different. There is not much hope though, looks like AMD did not backport the improvement to older architectures (unlike nVidia) - or maybe they're not physically capable, who knows.
As @thexai wrote, the PR code was reverted to master behavior to prepare for merge, so things will go back to normal for you in a new test build.

Android build failures are unrelated.

enen92 commented:
> I asked just in case your drivers were not up to date, then the result may be different. There is not much hope though, looks like AMD did not backport the improvement to older architectures (unlike nVidia) - or maybe they're not physically capable, who knows. As @thexai wrote, the PR code was reverted to master behavior to prepare for merge, so things will go back to normal for you in a new test build.
> 
> Android build failures are unrelated.

@CrystalP drivers were in fact outdated. However after the update the previous testbuild still gave the same result.

Tested an up to date revision of this patch and could not find any regressions, playback is flawless. Thanks!

thexai commented:
Jenkins build this please

enen92 commented:
jenkins build this please
